test/src/lib/backoff.spec.ts

```diff
@@ -0,0 +1,216 @@
+import { assert, expect } from 'chai';
+import * as sinon from 'sinon';
+
+import { StatusError } from '../../../src/lib/errors';
+import { withBackoff, OnFailureInfo } from '../../../src/lib/backoff';
+
+const DEFAULT_OPTIONS = {
+	maxRetries: 5,
+	maxDelay: 900000, // 15 minutes
+	minDelay: 10000, // 10 seconds
+};
+
+describe('lib/backoff', async () => {
+	let clock: sinon.SinonFakeTimers;
+
+	beforeEach(() => {
+		clock = sinon.useFakeTimers();
+	});
+
+	afterEach(() => {
+		clock.restore();
+	});
+
+	it('resolves after 3 retries', async () => {
+		// Create a function that will fail 3 times so it succeeds on the 4th
+		const failer = new Failer(3);
+		// Wrap function withBackoff
+		const fnWithBackoff = withBackoff(async () => {
+			return failer.willResolve('fails 3 times then resolves on 4th');
```

Comment:
Is not a deal breaker and I'm find to leave it as is, but I feel like the `Failer` class doesn't add much more than a stub would and it probably makes it a bit more difficult to read the code. 

As I said though I don't want to nitpick more on this as I think this works and we can always refactor in the future if needed.

Reply:
I don't like that I have to curry the function to make it work though so I'm interested in removing it if we can. I don't know how a stub could be made to fail N number of times

Reply:
right I see now. You'd just tick the clock and then make the stub resolve or reject at the callCount you want.